mautic/mautic
# Preference center unsubscribe toggle fix

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Y]
| New feature/enhancement? (use the a.x branch)      | [N]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [N]
| Automated tests included?              | [N] Only JS changes. No tests possible
| Related user documentation PR URL      | /
| Related developer documentation PR URL | /
| Issue(s) addressed                     | 

#### Description:

⚠️ this PR is based on https://github.com/mautic/mautic/pull/11442 as it is built on top of refactoring made there, so that PR should be merged first, this one rebased so it contains only commits relative to this bug fix.

The original problem was that the user cannot unsubscribe from a preference center that is configured like so:
<img width="878" alt="Screen Shot 2022-06-02 at 16 55 19" src="https://user-images.githubusercontent.com/1235442/171658333-38787df5-5b8d-43bb-bad5-0e5288cce1a5.png">

When the contacts want to unsubscribe from an email using this preference center then they cannot due to this JS error when they uncheck the channel they want to unsubscribe from:
```
ncaught TypeError: Cannot set properties of null (setting 'disabled')
    at togglePreferredChannel (prefcenter.js?vafdd094b:62:106)
    at HTMLInputElement.onclick (50a40ac2348acbe5a5de0e6a244253f13b2fd4357e801ee0313a104d0a03ae34:116:198)
```
There were 3 bugs hidden in this JS function:
1. The ID was wrong. It was missing the `lead_channels_` part. From what I was able to dig up it was changed in https://github.com/acquia/mc-cs/pull/517 which was already a fix for the missing part.
2. It failed even when I fixed the missing part because the 2 of the inputs that were suppose to be disabled were missing due to the **Show pause contact preferences** set to **No**. So we have to check if the inputs exist before disabling them.
3. The information whether the channel checkbox is checked or not was always false because there is [a hidden input added](https://github.com/mautic/mautic/commit/ab86dc0e299dbbe57dfcb9c31a6105d717b983e2) before the checkbox with the same ID and name for reason I don't yet comprehend.

I was checking that the similar JS function exists in other 2 places:
1. https://github.com/mautic/mautic/blob/4.x/app/bundles/EmailBundle/Views/Lead/preference_options.html.php#L15-L30
2. https://github.com/mautic/mautic/blob/4.x/app/bundles/LeadBundle/Assets/js/lead.js#L981-L999

~~And it will not work for the reasons above either. I asked @kuzmany to help me understand about the bug number 3. Awaiting reply. That's why I didn't jump into fixing the other 2 places as the bun number 3 may be reverted if @kuzmany answers that it was a mistake.~~ Discussed with @kuzmany and fixed the other 2 issues as well.

Docs: https://docs.acquia.com/campaign-studio/components/landing-pages/preference-center/

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Configure the preference center like the screenshot above.
3. Create a Landing Page as a preference center
4. The issue occurs when user add HTML block and add tag {channelfrequency} instead of adding preference center block and select channel frequency. Also the issue is reproducible with text block.
5. Create a segment email with the preference center configured and use the unsubscribe link in it. 
6. Send the email to a segment with your testing contact in it
7. Open the email
8. Click the unsubscribe link
9. Check that you cannot unsubscribe before this change
10. Check that you can unsubscribe after this change
11. Also check the preference center is not attached to the email you send. It displays different UI when you click on the unsubscribe link. So test that the checkboxes work there correctly.
12. Also check the preference center in the administration. Go to contacts > open one contact detail page > select preference center in the context menu and ensure it works.
13. Check that other variations of the preference center still work.



#### Other areas of Mautic that may be affected by the change:
1. Just the preference center in the public page. Not the one in the contact detail page.


[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. None as we have no infrastructure how to test JS code.
